src/commands/test_three_party_multi_user.rs

```diff
@@ -0,0 +1,158 @@
+//! `test-three-party-multi-user` - Scenario: Three parties, a manufacturer of prescription medicines,
+//! a purchaser of the medicines, and a supplier of the medicines. The manufacturer issues a ZSA
+//! for every dose of medicine produced. On purchase, the manufacturer transfers the corresponding
+//! number of ZSAs to the purchaser. The purchaser then transfers the ZSAs to the supplier, in
+//! exchange for the physical doses. The supplier burns the ZSAs after receiving them to signal the
+//! sale of the medicines.
+//!
+//! The tests ensure correct balance updates and transaction validity at each step of this scenario.
+
+use abscissa_core::{Command, Runnable};
+use orchard::issuance::compute_asset_desc_hash;
+use orchard::keys::Scope::External;
+
+use crate::commands::test_balances::{
+    print_balances, expected_balances_after_burn, expected_balances_after_transfer, BurnInfo,
+    TestBalances, TransferInfo, TxiBatch, check_balances_multi_user,
+};
+use crate::components::rpc_client::reqwest::ReqwestRpcClient;
+use crate::components::transactions::{
+    create_issue_transaction, mine_and_sync_others, sync_from_height,
+};
+use crate::components::user::User;
+use crate::prelude::*;
+
+/// Run the test scenario
+#[derive(clap::Parser, Command, Debug)]
+pub struct TestThreePartyMultiUserCmd {}
+
+impl Runnable for TestThreePartyMultiUserCmd {
+    /// Run the `test` subcommand.
+    fn run(&self) {
+        let config = APP.config();
+        let mut rpc_client = ReqwestRpcClient::new(config.network.node_url());
+
+        let num_users = 4; // Manufacturer, Purchaser, Supplier, Miner
+
+        let mut manufacturer = User::random(&config.wallet.miner_seed_phrase);
+        let mut purchaser = User::random(&config.wallet.miner_seed_phrase);
+        let mut supplier = User::random(&config.wallet.miner_seed_phrase);
+        let mut miner = User::random(&config.wallet.miner_seed_phrase);
+
+        manufacturer.reset();
+        purchaser.reset();
+        supplier.reset();
+        miner.reset();
+
+        let manufacturer_addr = manufacturer.address_for_account(0, External);
+
+        let mut wallets = [&mut manufacturer, &mut purchaser, &mut supplier, &mut miner];
+        let manufacturer_idx = 0;
+        let purchaser_idx = 1;
+        let supplier_idx = 2;
+        let miner_idx = 3;
```

Comment:
There is a lot of repetition here, could be improved, but as an initial proof-of-concept that it works I didn't get into it yet.